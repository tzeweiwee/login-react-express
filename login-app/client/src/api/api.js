import 'isomorphic-fetch';
import config from 'config';
import { checkStatus, getResponseJSON } from '../helpers/apiHelper';

function postAuthAPI (endpoint, params) {
  const { authenticationServiceURL } = config;
  return fetch(`${authenticationServiceURL}/${endpoint}`, {
      method: 'POST',
      headers: {
          "Accept": "application/json",
          "Content-type": "application/json"
      },
      body: JSON.stringify(params)
  });
}

const api = {
  postLogin: async (data) => {
    const result = await postAuthAPI('api/login', data);
    checkStatus(result);
    return getResponseJSON(result);
  }
}

export default api;