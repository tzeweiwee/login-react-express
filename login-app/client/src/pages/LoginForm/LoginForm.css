.errorMessage {

}

.loginContainer {
  width: 500px;
  height: 1000px;
}