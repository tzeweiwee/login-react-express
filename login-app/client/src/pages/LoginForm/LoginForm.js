import React, { Component } from "react";
import InputField from "../../components/InputField/InputField";
import Button from "../../components/Button/Button";
import styles from "./LoginForm.css";
import api from "../../api/api";

export default class LoginForm extends Component {
  constructor(props) {
    super(props);
    this.handleEmail = this.handleEmail.bind(this);
    this.handlePassword = this.handlePassword.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.state = {
      email: "",
      password: "",
      errors: [],
      success: false
    };
  }

  handleEmail = e => {
    this.setState({
      email: e.target.value
    });
  };

  handlePassword(e) {
    this.setState({
      password: e.target.value
    });
  }

  validate(email, password) {
    const errors = [];
    if (email.length === 0) {
      errors.push("Email is required");
    } else {
      if (
        email.split("").filter(x => x === "@").length !== 1 ||
        email.indexOf(".") === -1
      ) {
        errors.push("Invalid Email");
      }
    }

    if (password.length === 0) {
      errors.push("Password is required");
    }

    return errors;
  }

  handleSubmit = async e => {
    e.preventDefault();
    const { email, password } = this.state;
    const errors = this.validate(email, password);
    this.setState({
      errors,
      success: false
    });
    if (!errors.length) {
      try {
        let result = await api.postLogin({ email, password });
        this.setState({
          success: result.body
        });
      } catch (err) {
        this.setState({
          errors: [err.message]
        });
      }
    }
  };

  render() {
    const { email, password, errors, success } = this.state;
    let errorMessages;
    if (errors.length) {
      errorMessages = errors.map(err => <p key={err}>{err}</p>);
    }
    return (
      <div className={styles.loginContainer}>
        <form>
          <InputField
            onChange={this.handleEmail}
            type="email"
            value={email}
            labelText="Email"
            name="email"
            placeholder="Email"
          />
          <InputField
            onChange={this.handlePassword}
            type="password"
            value={password}
            labelText="Password"
            name="password"
            placeholder="Password"
          />
          {errorMessages && (
            <label className={styles.errorMessage}>{errorMessages}</label>
          )}
          {success && (
            <label className={styles.successMessage}>{success}</label>
          )}
          <Button onClick={this.handleSubmit} type="submit" text="Sign In" />
        </form>
      </div>
    );
  }
}
