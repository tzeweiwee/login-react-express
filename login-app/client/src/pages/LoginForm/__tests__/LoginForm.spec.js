import React from "react";
import { shallow } from "enzyme";
import LoginForm from "../LoginForm";
import api from "../../../api/api";

const emailError = ["Invalid Email"];
const passwordError = ["Password is required"];

const event = {
  preventDefault: () => {}
};

describe("LoginForm", () => {
  it("should contain validation errors when email is invalid", () => {
    const wrapper = shallow(<LoginForm />);
    const userInput = { email: "invalid", password: "abc" };
    wrapper.setState(userInput);
    const errors = wrapper
      .instance()
      .validate(userInput.email, userInput.password);
    expect(errors).toEqual(expect.arrayContaining(emailError));
  });

  it("should contain validation errors when password is missing", () => {
    const wrapper = shallow(<LoginForm />);
    const userInput = { email: "test@test.com", password: "" };
    wrapper.setState(userInput);
    const errors = wrapper
      .instance()
      .validate(userInput.email, userInput.password);
    expect(errors).toEqual(expect.arrayContaining(passwordError));
  });

  it("should call login API when there are no validation errors", async () => {
    const wrapper = shallow(<LoginForm />);
    const userInput = { email: "test@test.com", password: "1234" };
    const apiSpy = jest.spyOn(api, "postLogin");
    wrapper.setState(userInput);
    await wrapper.instance().handleSubmit(event);

    expect(apiSpy).toHaveBeenCalled();
  });

  it("should match LoginForm snapshot", () => {
    const wrapper = shallow(<LoginForm />);
    expect(wrapper).toMatchSnapshot();
  });
});
