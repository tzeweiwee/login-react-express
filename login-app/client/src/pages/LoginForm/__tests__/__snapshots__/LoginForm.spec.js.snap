// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match LoginForm snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form>
        <InputField
                labelText="Email"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="email"
                value=""
        />
        <InputField
                labelText="Password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
        />
        <Button
                onClick={[Function]}
                text="Sign In"
                type="submit"
        />
</form>,
      "className": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <InputField
            labelText="Email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
/>,
          <InputField
            labelText="Password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
/>,
          undefined,
          false,
          <Button
            onClick={[Function]}
            text="Sign In"
            type="submit"
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "labelText": "Email",
            "name": "email",
            "onChange": [Function],
            "placeholder": "Email",
            "type": "email",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "labelText": "Password",
            "name": "password",
            "onChange": [Function],
            "placeholder": "Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        undefined,
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onClick": [Function],
            "text": "Sign In",
            "type": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form>
          <InputField
                    labelText="Email"
                    name="email"
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value=""
          />
          <InputField
                    labelText="Password"
                    name="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
          />
          <Button
                    onClick={[Function]}
                    text="Sign In"
                    type="submit"
          />
</form>,
        "className": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <InputField
              labelText="Email"
              name="email"
              onChange={[Function]}
              placeholder="Email"
              type="email"
              value=""
/>,
            <InputField
              labelText="Password"
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
/>,
            undefined,
            false,
            <Button
              onClick={[Function]}
              text="Sign In"
              type="submit"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "labelText": "Email",
              "name": "email",
              "onChange": [Function],
              "placeholder": "Email",
              "type": "email",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "labelText": "Password",
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          undefined,
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClick": [Function],
              "text": "Sign In",
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
